### PR TITLE
nixos/nextcloud: support SSE-C for S3 primary storage

### PR DESCRIPTION
**Pull request by tensor5**
Add configuration option to enable [server-side encryption with customer-provided keys][1] (SSE-C) when using S3 as primary storage in Nextcloud.

[1]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 1]
```diff
@@ -514,6 +514,30 @@ in {
               `http://hostname.domain/bucket` instead.
             '';
           };
+          useSseC = mkOption {
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-apps/nextcloud.nix:517]: This option should not exist, rather use whether sseCKeyFile is unset or set

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 2]
```diff
@@ -266,3 +266,5 @@ In addition to numerous new and upgraded packages, this release has the followin
 - The option `services.nomad.extraSettingsPlugins` has been fixed to allow more than one plugin in the path.
 
 - The option `services.prometheus.exporters.pihole.interval` does not exist anymore and has been removed.
+
+- `nextcloud` has an option to enable SSE-C in S3.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2305.section.md:270]: Please see the comment right under the section heading

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 3]
```diff
@@ -514,6 +514,30 @@ in {
...
+              Use [server-side encryption with customer-provided keys][1].
+
+              A random 32-byte key must be provided as a base64 string, e.g.
+              with the command
+
+              ```
+              openssl rand 32 | base64
+              ```
+
+              [1]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/nextcloud.nix:530]:
```suggestion
              Use [server-side encryption with customer-provided keys][1].

              A random 32-byte key must be provided as a base64 string.
              One can be generated with the following command:
              ```
              openssl rand 32 | base64
              ```

              [1]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html
```

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 4]
```diff
@@ -514,6 +514,30 @@ in {
...
+            '';
+          };
+          sseCKeyFile = mkOption {
+            type = types.str;
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/nextcloud.nix:534]:
```suggestion
            type = types.nullOr types.path;
```
or where secret files just types.str?

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 1 above)
Comment (tensor5) [reply nixos/modules/services/web-apps/nextcloud.nix]: I agree

https://github.com/NixOS/nixpkgs/blob/8008e73c3878ec4b4e9633c89c5113c1cfc6abbe/nixos/modules/services/web-apps/nextcloud.nix#L517

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (tensor5) [reply nixos/modules/services/web-apps/nextcloud.nix]: https://github.com/NixOS/nixpkgs/blob/8008e73c3878ec4b4e9633c89c5113c1cfc6abbe/nixos/modules/services/web-apps/nextcloud.nix#L518

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 3 above)
Comment (tensor5) [reply nixos/modules/services/web-apps/nextcloud.nix]: Thank you. Please check the new description after the removal of `useSseC` [here](https://github.com/NixOS/nixpkgs/blob/8008e73c3878ec4b4e9633c89c5113c1cfc6abbe/nixos/modules/services/web-apps/nextcloud.nix#L521).

File: nixos/doc/manual/release-notes/rl-2305.section.md (hunk 2 above)
Comment (tensor5) [reply nixos/doc/manual/release-notes/rl-2305.section.md]: Done

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please squash your commits.

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
@ofborg test nextcloud

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**